### PR TITLE
Version 0.18.0

### PR DESCRIPTION
**Pull request by karosis88**
# Changelog

## 0.18.0

- Add support for HTTPS proxies. (#745, # 786)
- Handle `sni_hostname` extension with SOCKS proxy. (#774)
- Change the type of `Extensions` from `Mapping[Str, Any]` to `MutableMapping[Str, Any]`. (#762)
- Handle HTTP/1.1 half-closed connections gracefully. (#641)
- Drop Python 3.7 support. (#727)

## TODO

- [ ] Update release date

### CODE REVIEW

**Comment by karosis88** [comment]
It appears we can't release this without the https://github.com/encode/httpcore/pull/757

**Comment by tomchristie** [comment]
> It appears we can't release this without the https://github.com/encode/httpcore/pull/757

Could you explain what you mean here?

File: httpcore/__init__.py
Comment (karosis88) [inline httpcore/__init__.py:133]:
```suggestion
__version__ = "0.18.0"
```

File: CHANGELOG.md
Comment (karosis88) [inline CHANGELOG.md:7]:
```suggestion
## 0.18.0
```

**Comment by trim21** [comment]
according to https://github.com/encode/httpx/pull/2801#issuecomment-1669347112, the todo list also include https://github.com/encode/httpcore/pull/757

**Comment by tomchristie** [comment]
Okay, here's how our `CHANGELOG` for this version looks at this point in time...

- Add support for HTTPS proxies. (#745, # 786)
- Handle `sni_hostname` extension with SOCKS proxy. (#774)
- Change the type of `Extensions` from `Mapping[Str, Any]` to `MutableMapping[Str, Any]`. (#762)
- Handle HTTP/1.1 half-closed connections gracefully. (#641)
- Drop Python 3.7 support. (#727)

The most substantial other bits of work that we have on the table right now are...

- Some remaining issues around async cancellations. (#756, #785)
- Explicit proxy configuration between forwarding/connect. (#735)

I'd probably suggest that we press on with a 0.18.0 release as we currently stand...

- The async cancellation issues are quite involved, and I don't want us to be time pressured in resolving them.
- This will give us time for our users to test our HTTPS proxy support before we add any further functionality there.

There's also already plenty more for us to wrap in an `httpx` 0.25 release - https://github.com/encode/httpx/pull/2801 once we've got this `httpcore` release done.

Thoughts, team?

**Comment by karosis88** [comment]
There is already an easy solution in #757  for our async cancellations. I suggest that we include that simple modification, but if you don't like it, I am fine with this changelog.

**Comment by tomchristie** [comment]
> I am fine with this changelog.

This would be my preference yes.

To my eyes #757 is too awkwardly gnarly to just wave on through, I'd like us to slow down on that and get it comprehensively correct.

**Comment by karosis88** [comment]
In addition, I will gladly include #734 to this release. It appears to be a part of #722.

**Comment by T-256** [comment]
> In addition, I will gladly include #734 to this release. It appears to be a part of #722.

IMO they are separated from each other, But yes, I agree to include it in 0.18.0

**Comment by tomchristie** [comment]
> @[karosis88](https://github.com/karosis88) karosis88 requested a review from [tomchristie](https://github.com/tomchristie)

I'm happy to approve this, once the description matches the CHANGELOG.

**Comment by karosis88** [comment]
I will add a release date after your review.

**Review by tomchristie — APPROVED** [review]

**Comment by tomchristie** [comment]
Let's get a release draft ready before we merge this...

**Comment by karosis88** [comment]
> Let's get a release draft ready before we merge this...

Can i create a release draft?

**Comment by tomchristie** [comment]
> Can I draft a new release?

I don't recall if you're able to do that with the `@encode/maintainers` permissions.
 
In any case I've created one here... https://github.com/encode/httpcore/releases/tag/untagged-71ea7b9ccfc1bfebd276

File: CHANGELOG.md
Comment (karosis88) [inline CHANGELOG.md:7]:
```suggestion
## 0.18.0  (September 8th, 2023)
```

File: CHANGELOG.md
Comment (T-256) [inline CHANGELOG.md:9]:
```suggestion
- Add support for HTTPS proxies. (#745, #786)
```
it is also on description

Comment (karosis88) [reply CHANGELOG.md]: Good catch, thanks!

**Pull request merged**
